### ISSUE TITLE
Finding the source line and column number for AST nodes?

### ISSUE THREAD

**Issue by plevold** [lfortran/lfortran#1036]
I'm looking into using the `--show-ast` feature of LFortran to analyze some of my Fortran source files (hopefully it will turn into something useful that can be published at some point). Would it be possible to have LFortran write information needed to determine the corresponding line/column number in the source file that an AST node originates from?

I did have a look at the source code and found some structs named `Location` and `locationManager`, but my experience with writing compilers are totally non-existent so I couldn't figure out how it's used.

For example, the code

```fortran
subroutine sub(i)
    integer, intent(in) :: i
    
    print *, i
end subroutine
```

Give me the following syntax tree:

```
(TranslationUnit 
    [(Subroutine 
        sub 
        [(i)] 
        [] 
        () 
        () 
        [] 
        [] 
        [] 
        [(Declaration 
            (AttrType 
                TypeInteger 
                [] 
                () 
                None
            ) 
            [(AttrIntent 
                In
            )] 
            [(i 
            [] 
            [] 
            () 
            None 
            ())] 
            (TriviaNode 
                [] 
                [(EndOfLine)
                (EndOfLine)]
            )
        )] 
        [(Print 
            0 
            () 
            [i] 
            ()
        )] 
        []
    )]
)
```

I am looking for a way to determine that `Print` originates from line 4, columns 4 to 14 in the original source file.

**Comment by certik** [comment]
Each node as a Location structure that has first and last index. This linear index can then be remapped into a filename+line+column using the LocationManager.

Depending on what you are trying to do, we could save the AST + location into some kind of a JSON so that it's easy to import into, say, Python.

**Comment by plevold** [comment]
> Depending on what you are trying to do, we could save the AST + location into some kind of a JSON so that it's easy to import into, say, Python.

Yes that would be exactly what I'm looking for! Is this something you plan on implementing anyways? I could have a look at it myself, but given my limited time for such projects, limited C++ experience and total lack of compiler dev experience I suspect it would take ages to get anything done 🤔

**Comment by certik** [comment]
I can implement it for you. But you then need to find time to at least try it. :)

**Comment by plevold** [comment]
Wow, thanks! I'll for sure find time for testing!

**Comment by michaelkonecny** [comment]
I would love to have this as well! (Will find time to test.)

I want to write a script will check all my files and print out all the locations where I forgot to use double precision for inline values 
(i.e. `a = b / 2.0` or worse `a = b / 2` instead of `a = b / 2d0`).
Not sure how this works in other compilers, but it causes a loss of precision in ifort on Windows and I literally spend days for every project finding these.

If I could get a JSON where I could look at the precision for each value and have the location in the source file, that would be just terrific!

**Comment by certik** [comment]
Sorry I forgot to update this issue. We already implemented JSON printing:

* https://github.com/lfortran/lfortran/pull/1071
* https://github.com/lfortran/lfortran/pull/1078

It prints location information, but as a linear index. We still need to either save the line/column/file, or give you information how to do this yourself. @michaelkonecny or @plevold maybe you can implement this for us.

@michaelkonecny your idea should be a regular LFortran warning (configurable via a compile time option). But you can play with the JSON and give us feedback. I think you need ASR for this, not AST, since you need to know the types.

**Comment by michaelkonecny** [comment]
Ifort has the `-fpconstant` flag, but it doesn't cover all cases.
For example, in
```
real(8) :: a
a = 1.0/3.0
```
the value stored to `a` is still only calculated with single precision.
The above is a good example of what I want to discover automatically with a script - it happens a lot (we often rewrite Matlab code to Fortran, one starts by copy pasting and it's easy to miss `1e7` in place of `1d7`).

Can you give me a hint on how to convert the linear index to file/line/column?
I would be up to implementing this.

Thanks

**Comment by certik** [comment]
For now you can use this function: https://github.com/lfortran/lfortran/blob/d2f947eed49002391830291420be48bcb74c767b/src/libasr/codegen/asr_to_llvm.cpp#L375, although I don't like how it is implemented there, it should accept the LocationManager as an argument, and not read the file at all (@Thirumalai-Shaktivel, we should fix that).  The other place where this is done is when printing error messages here: https://github.com/lfortran/lfortran/blob/d2f947eed49002391830291420be48bcb74c767b/src/libasr/diagnostics.cpp#L121, just call `LocationManager::pos_to_linecol`.

**Comment by michaelkonecny** [comment]
OK, I will look into that!

**Comment by michaelkonecny** [comment]
Do we want the line/column information only in the JSON or potentially everywhere?
If only in JSON, I can keep add a pointer to LocationManager in JsonBaseVisitor and initialize it in its constructor.
If everywhere, I would pass LocationManager to all the visitors and their visit functions, which is a bit more work and makes things a bit less readable.
Personally, I would prefer the first option.

**Comment by certik** [comment]
Let's start with just JSON, with some minimal clean approach and then let's iterate.

I would recommend storing just the linear (integer first/last) location information and then store the information from the LocationManager. Otherwise you have to store line, column and file (number) for each node, which I think will get quite verbose. Typically one would store a list of files first anyway, and then reference them with an index. So we might as well just use the linear index, and store filenames, line ends, etc. in the "header", and then we can provide simple Python functions to convert linear index into line+column+filename.

**Comment by michaelkonecny** [comment]
Hmm... I just finished implementing it the verbose way...
But what you're saying makes sense, I didn't think of that... I'll look into it after the holidays. I think converting the linear index to line+column+file is pretty straight-forward, no need to provide a Python function to do it (where would you put it?). I would just describe the linear index in the `--json` switch documentation.

What do you think about storing the pointer to LocationManager and setting it in the constructor like this:
```cpp
template <class Struct>
class JsonBaseVisitor : public BaseVisitor<Struct>
{
private:
    Struct& self() { return static_cast<Struct&>(*this); }
public:
    std::string s, indtd = "";
    int indent_level = 0, indent_spaces = 4;
    LocationManager* lm = nullptr;
public:
    JsonBaseVisitor(LocationManager &lmref) {
        lm = &lmref;
        s.reserve(100000);
    }
// ...
```
?

In terms of readability, I think it's better than the alternative - passing `lm` to all the Visitors and their visit functions, but I also don't like it, because the lifetime of a JsonBaseVisitor can be different than that of an AST passed to visit_ast().
If JsonBaseVisitor ever gets reused for a different AST, it could happen to be using an non-relevant LocationManager, which would be an annoying thing to debug.
I'm not sure which is better.

**Comment by certik** [comment]
Yes, these lifetimes are the usual problem with C++. There is no good solution, one can use shared pointer, but then one has to initialize LocationManager using it as well. All things considered, I just pass it in via a constructor like you did. I typically use `&` and initialize it via the constructor initializer.

If you are going to go with storing just the linear index, you don't need to store the location manager in the class at all.

Regarding the remapping, it's not complicated, but it's not completely trivial either, you have to remap via the intervals (as the source code changes), and you have to do it twice, once for a C preprocessor, once for prescanning. We can put it into some examples and test it by the CI. Then people can just take it and use in their Python projects.

**Comment by michaelkonecny** [comment]
> Regarding the remapping, it's not complicated, but it's not completely trivial either, you have to remap via the intervals (as the source code changes), and you have to do it twice, once for a C preprocessor, once for prescanning. We can put it into some examples and test it by the CI. Then people can just take it and use in their Python projects.

I see. I thought it was simpler. If this is the case then I have two questions and a point:
1) is the added verbosity really a problem? (personally I could probably live with it)
2) if it is, can't we do something simpler, like just add line+col+fileID and then have a list of files and their IDs at the end (as the filename is susceptible be the most verbose part)
3) if the overall output size is the issue, maybe we could start by minifying the JSON, this would more than counteract the added verbosity.

If possible, I think it would be a good idea to make the JSON as self-explanatory as possible. Having to use an extra script to decode the location means people will be less likely to use it, IMO.

**Comment by certik** [comment]
Yes, we can do any of that. I think it depends on the application which option would be the best. I agree it's not ideal to use an extra script to decode. One issue that I will point out is that if you decode ahead of time, you lose the ability to track include files and macro expansions, you will just get the filename + line/column of the final destination, but it can be a different file. 

To make some progress, why don't you do the remapping and save line/column/filename (either a string or an id), and that would be a good start. Then we can iterate upon this later if needed.

**Comment by michaelkonecny** [comment]
About the LocationManager pointer in JsonBaseVisitor:
I've tried thinking about it from left and right and I can't think of a better solution. I looked more into passing a LM reference to all the visit functions, but that would mean having to create LocationManager in a lot of places where it's just not needed.

Here's a PR: https://github.com/lfortran/lfortran/pull/1142 with the implementation with the pointer.
For now I kept the previous linear location as well. Should it be removed?

I also found out `test_ast_to_json` is probably failing now, so I guess that should be fixed. It also doesn't seem to be connected to CI.
Could you provide any pointers on how to build/run the test and do we want to connect it to CI?

Thanks.

**Comment by certik** [comment]
Yes, I think we do not test the JSON output yet. We should.

Probably the easiest way is to hook it up into `./run_tests.py`, just like AST "lisp" like output.

**Comment by michaelkonecny** [comment]
I see. OK, that seems straight-forward.

In that case `test_ast_to_json.cpp` will still have no use. Shouldn't it be deleted?
I didn't figure out how this test was supposed to be run, perhaps someone can enlighten me (I figure it's an option I need to pass to cmake or make, but I'm not sure how to find out which one and I'm a bit lazy...).

**Comment by michaelkonecny** [comment]
On a related note, I added a mention of the `--json` option to the documentation here: #1143

**Comment by certik** [comment]
I need to look into `test_ast_to_json.cpp`, it might be some old code that we can remove. I would worry about it later.

**Comment by michaelkonecny** [comment]
Just found out today that the JSON isn't valid on Windows because the filenames contain backslashes and they don't get escaped.
A quick fix would be simple - just escape the backslashes, but a better solution would probably be to use a library like [RapidJSON](https://rapidjson.org/) to generate the JSON.

This would also allow to generate a non-indented ("minified") version of the JSON in order to reduce the output size.
I guess this should happen when a switch like `--json-minify` is present.

What do you say?

The library is header-only, which is nice.
Does you guys use [vcpkg](https://github.com/Microsoft/vcpkg/) by any chance?

**Comment by michaelkonecny** [comment]
@certik I started implementing this using RapidJSON, so stop me now or never ;)

**Comment by certik** [comment]
RapidJSON is good, we already have support for it: https://github.com/lfortran/lfortran/blob/ea5234f19dfcf286177fdf51af2e2438c78478ed/CMakeLists.txt#L210, see here how to use it: https://github.com/lfortran/lfortran/blob/ea5234f19dfcf286177fdf51af2e2438c78478ed/src/lfortran/ast_to_json.cpp.

**Comment by certik** [comment]
We use Conda mostly for dependencies.

**Comment by certik** [comment]
Regarding performance, I am not sure if RapidJSON is faster than just emitting the string as we do. I would suspect it is not faster. Once you have something working, we can benchmark it. If RapidJSON is slower, then we should use our existing approach.

I think RapidJSON is faster for reading JSON than something we can done by hand.

**Comment by michaelkonecny** [comment]
I would also expect it to be slower, but probably not by much. It seems to be quite cleverly designed.

I wanted to use it mainly for escaping the special characters in file paths. This could of course also be implemented on our own, but why bother when there's a working library.

Similarly for the minifying/prettifying (although this is even simpler to do than the escaping).

I'm almost done, so we can benchmark it and see...

**Comment by michaelkonecny** [comment]
I tested it on a 60 kB (~1500 lines) source file.
With RapidJSON, it's about 5 % slower than just emmiting the string, which is less than I expected.
(I measured the whole `lfortran file.f90 --show-ast --json` run, so that time includes parsing the fortran file and everything else.)

Is that acceptable?

**Comment by certik** [comment]
Can you post your code somewhere? Let me benchmark it myself. One has to be very careful with benchmarking, such as compiling LFortran in release mode, disabling asserts, etc. Also it's good to benchmark just the backend itself.

**Comment by plevold** [comment]
@certik @michaelkonecny I had a look at the JSON format and this looks super nice! It's exactly what I was looking for. Thanks a lot!

One thing I'd like to note with regards to the line/column numbering is however that Unicode characters can cause problems with determining the column number of a node. Is Unicode characters except for the ASCII ones allowed in Fortran source files? I'm actually not sure, but most compilers including LFortran accepts string literals with Unicode in them. 

For example, which column numbers does the following three string literals cover?

```fortran
program  main
    implicit none

    write(*,*) 'a'
    write(*,*) '🤩'
    write(*,*) 'a̪̘͇̅'
end program
```

Currently, I get 
1. Columns 16 to 18
2. Columns 16 to 21
3. Columns 16 to 26

The line/column indicator in VSCode on the other hand says the following:
1. Columns 16 to 18
2. Columns 16 to 18
3. Columns 16 to 22

If you look at it manually when using a monospace font at I would "guess" the following
1. Columns 16 to 18
2. Columns 16 to 19 (the emoji is approximately two "regular" characters wide)
3. Columns 16 to 18

Why does this happen? Because determining the width of a set of bytes from a Unicode encoded string is surprisingly difficult. 🤩 for example (U+1F929 : GRINNING FACE WITH STAR EYES) is 4 bytes long in UTF-8, but usually approximately two ASCII characters wide. With combining characters it gets even more tricky. The ridiculous `a̪̘͇̅` is the same width as a regular `a`, but is actually 5 Unicode codepoints totalling 9 bytes in length:
```
U+0061 : LATIN SMALL LETTER A
U+0305 : COMBINING OVERLINE {overscore, vinculum}
U+032A : COMBINING BRIDGE BELOW
U+0318 : COMBINING LEFT TACK BELOW
U+0347 : COMBINING EQUALS SIGN BELOW
```

Why is this important? For example, if I add an error on the last line the indicator in the LFortran error message will be offset from the actual text it tries to point to:
![image](https://user-images.githubusercontent.com/10157556/210874357-8ae699c4-2db8-48e0-9587-4f9dc284aa48.png)

Now for 99.9999% of all Fortran code out there it is probably reasonable to assume that only ASCII characters have been used which will give correct answers with the current calculation. I'm not really sure if it's worth pursuing a fix for this. Even VSCode seems to struggle with combining characters. Also, in the context of AST serialization if somebody actually needs this kind of precision then the `first` and `last` indicators can still be used in combination with a Unicode segmentation algorithm at the other end.

If somebody actually want to go even further down this rabbit hole I'll note that the Rust compiler actually seems to handle this quite well. It could be worth trying to locate their code for this. For example, with the following erroneous piece of code the compiler is able to pinpoint the columns spanning the error:

```rust
fn main() {
    println!("Hello world🤩a̪̅"); error
}
``` 

Compiler output:
![image](https://user-images.githubusercontent.com/10157556/210877546-28002a02-49d7-4764-b07a-3c9026765616.png)

**Comment by plevold** [comment]
> The line should always be correct, and the column is in bytes.

Yes, that will work.

> To convert bytes to unicode "code points" (which I think is what is needed above to get correct error messages), I think all we need is to take the input string and column in bytes, truncate the input line to end exactly at the column, then use this function:

Well, it depends on what one wants to achieve. VSCode counts one code point as a column, but to align text in a terminal that use a monospace font this will not work. The string `🤩a̪̘͇̅` is 6 Unicode codepoints, but I need approximately 3 dots (or spaces) to align with the `X` behind it:

```
🤩a̪̘͇̅X
...^
```

To calculate this properly one would need to not only identify the codepoints, but the different grapheme clusters (which is 2 in this case) and then estimate their width. According to [this](https://www.reddit.com/r/rust/comments/gpw2ra/how_is_the_rust_compiler_able_to_tell_the_visible/) post, the Rust compiler use [this](https://github.com/unicode-rs/unicode-width) library to do these calculations. 

But as you say, this is best left for the consumer of the serialized JSON to handle as computing this for every node would be expensive.

**Comment by michaelkonecny** [comment]
@certik 
> Can you post your code somewhere? Let me benchmark it myself. One has to be very careful with benchmarking, such as compiling LFortran in release mode, disabling asserts, etc. Also it's good to benchmark just the backend itself.

Good point. I forgot about that.

You can check out the branch [1036-json-line-number-WIP](https://github.com/michaelkonecny/lfortran/tree/1036-json-line-number-WIP) from my fork [michaelkonecny/lfortran](https://github.com/michaelkonecny/lfortran).

Please keep in mind:
- ASR doesn't work yet (the overloads in ASRJsonVisitor in pickle.cpp need to be finished)
- I haven't figured out how to connect RapidJSON with conda and CMake properly, so I put it in `src/` for now. 
I might need some help with that.
- The code might need some refactoring/renaming, but principally I think it's fine
- "minified" JSON is output as default. Uncomment rapidjson::PrettyWriter in ast.h to get prettified JSON (also `#include <rapidjson/prettywriter.h>`). I would like to connect it to the `--indent` switch later.

Please keep me posted so I know whether I should finish it.

**Comment by certik** [comment]
@plevold good point, and indeed the grapheme clusters computation I think is very complex. The Rust library seems to just use a lookup table, and it seems simpler than I thought, I was under the impression that there are some languages that have arbitrarily complex grapheme clusters and it's complicated how to determine character widths. Either way, I think leaving it to the user is the best way forward at this point. Maybe later we can revisit this.

@michaelkonecny thanks, I'll try it soon.

**Comment by certik** [comment]
@michaelkonecny sure. I can't compile your branch, I get:
```console
$ make
Consolidate compiler generated dependencies of target asr
[  0%] Building CXX object src/libasr/CMakeFiles/asr.dir/codegen/asr_to_cpp.cpp.o
In file included from /Users/ondrej/repos/lfortran/lfortran/src/libasr/codegen/asr_to_cpp.cpp:4:
/Users/ondrej/repos/lfortran/lfortran/src/libasr/../libasr/asr.h:20348:33: error: expected ';' after expression
        global_scope.SetObject()
                                ^
                                ;
```

**Comment by michaelkonecny** [comment]
My apologies, my bad...
I created a new branch here: [1036-json-line-number-WIP2](https://github.com/michaelkonecny/lfortran/tree/1036-json-line-number-WIP2) (mind the '2' at the end).

Do you think you could pull it and then do:
```
./build0.sh
python src/libasr/asdl_cpp_ast.py grammar/AST.asdl src/lfortran/ast.h
make -j8
```
?

(Explanation: The RapidJSON implementation is only finished for AST so far. I reverted `asdl_cpp.py` to the previous state (without RapidJSON) and added `asdl_cpp_ast.py` just for AST just for the purpose of the benchmark.)

Thank you.
